File: rust/src/api/endpoint.rs
// this file is @generated
use crate::{
    error::Result,
    models::*,
    Configuration,
};

#[derive(Default)]
pub struct EndpointListOptions {
    /// Limit the number of returned items
    pub limit: Option<i32>,

    /// The iterator returned from a prior invocation
    pub iterator: Option<String>,

    /// The sorting order of the returned items
    pub order: Option<Ordering>,
}

#[derive(Default)]
pub struct EndpointCreateOptions {
    pub idempotency_key: Option<String>,
}

#[derive(Default)]
pub struct EndpointRecoverOptions {
    pub idempotency_key: Option<String>,
}

#[derive(Default)]
pub struct EndpointReplayMissingOptions {
    pub idempotency_key: Option<String>,
}

#[derive(Default)]
pub struct EndpointRotateSecretOptions {
    pub idempotency_key: Option<String>,
}

#[derive(Default)]
pub struct EndpointSendExampleOptions {
    pub idempotency_key: Option<String>,
}

#[derive(Default)]
pub struct EndpointGetStatsOptions {
    /// Filter the range to data starting from this date.
    ///
    /// RFC3339 date string.
    pub since: Option<String>,

    /// Filter the range to data ending by this date.
    ///
    /// RFC3339 date string.
    pub until: Option<String>,
}

pub struct Endpoint<'a> {
    cfg: &'a Configuration,
}

impl<'a> Endpoint<'a> {
    pub(super) fn new(cfg: &'a Configuration) -> Self {
        Self {
            cfg,
        }
    }

    /// List the application's endpoints.
    pub async fn list(
        &self,
        app_id: String,
        options: Option<EndpointListOptions>,
    ) -> Result<ListResponseEndpointOut> {
        let EndpointListOptions {
            limit,
            iterator,
            order,
        } = options.unwrap_or_default();

        crate::request::Request::new(
            http1::Method::GET,
            "/api/v1/app/{app_id}/endpoint",
        )
        .with_path_param(
            "app_id", app_id,
        )
        .with_optional_query_param(
            "limit", limit,
        )
        .with_optional_query_param(
            "iterator", iterator,
        )
        .with_optional_query_param(
            "order", order,
        )
        .execute(self.cfg)
        .await
    }

    /// Create a new endpoint for the application.
    ///
    /// When `secret` is `null` the secret is automatically generated
    /// (recommended).
    pub async fn create(
        &self,
        app_id: String,
        endpoint_in: EndpointIn,
        options: Option<EndpointCreateOptions>,
    ) -> Result<EndpointOut> {
        let EndpointCreateOptions {
            idempotency_key,
        } = options.unwrap_or_default();

        crate::request::Request::new(
            http1::Method::POST,
            "/api/v1/app/{app_id}/endpoint",
        )
        .with_path_param(
            "app_id", app_id,
        )
        .with_optional_header_param(
            "idempotency-key",
            idempotency_key,
        )
        .with_body_param(endpoint_in)
        .execute(self.cfg)
        .await
    }

    /// Get an endpoint.
    pub async fn get(
        &self,
        app_id: String,
        endpoint_id: String,
    ) -> Result<EndpointOut> {
        crate::request::Request::new(
            http1::Method::GET,
            "/api/v1/app/{app_id}/endpoint/{endpoint_id}",
        )
        .with_path_param(
            "app_id", app_id,
        )
        .with_path_param(
            "endpoint_id",
            endpoint_id,
        )
        .execute(self.cfg)
        .await
    }

    /// Update an endpoint.
    pub async fn update(
        &self,
        app_id: String,
        endpoint_id: String,
        endpoint_update: EndpointUpdate,
    ) -> Result<EndpointOut> {
        crate::request::Request::new(
            http1::Method::PUT,
            "/api/v1/app/{app_id}/endpoint/{endpoint_id}",
        )
        .with_path_param(
            "app_id", app_id,
        )
        .with_path_param(
            "endpoint_id",
            endpoint_id,
        )
        .with_body_param(endpoint_update)
        .execute(self.cfg)
        .await
    }

    /// Delete an endpoint.
    pub async fn delete(
        &self,
        app_id: String,
        endpoint_id: String,
    ) -> Result<()> {
        crate::request::Request::new(
            http1::Method::DELETE,
            "/api/v1/app/{app_id}/endpoint/{endpoint_id}",
        )
        .with_path_param(
            "app_id", app_id,
        )
        .with_path_param(
            "endpoint_id",
            endpoint_id,
        )
        .returns_nothing()
        .execute(self.cfg)
        .await
    }

    /// Partially update an endpoint.
    pub async fn patch(
        &self,
        app_id: String,
        endpoint_id: String,
        endpoint_patch: EndpointPatch,
    ) -> Result<EndpointOut> {
        crate::request::Request::new(
            http1::Method::PATCH,
            "/api/v1/app/{app_id}/endpoint/{endpoint_id}",
        )
        .with_path_param(
            "app_id", app_id,
        )
        .with_path_param(
            "endpoint_id",
            endpoint_id,
        )
        .with_body_param(endpoint_patch)
        .execute(self.cfg)
        .await
    }

    /// Get the additional headers to be sent with the webhook.
    pub async fn get_headers(
        &self,
        app_id: String,
        endpoint_id: String,
    ) -> Result<EndpointHeadersOut> {
        crate::request::Request::new(
            http1::Method::GET,
            "/api/v1/app/{app_id}/endpoint/{endpoint_id}/headers",
        )
        .with_path_param(
            "app_id", app_id,
        )
        .with_path_param(
            "endpoint_id",
            endpoint_id,
        )
        .execute(self.cfg)
        .await
    }

    /// Set the additional headers to be sent with the webhook.
    pub async fn update_headers(
        &self,
        app_id: String,
        endpoint_id: String,
        endpoint_headers_in: EndpointHeadersIn,
    ) -> Result<()> {
        crate::request::Request::new(
            http1::Method::PUT,
            "/api/v1/app/{app_id}/endpoint/{endpoint_id}/headers",
        )
        .with_path_param(
            "app_id", app_id,
        )
        .with_path_param(
            "endpoint_id",
            endpoint_id,
        )
        .with_body_param(endpoint_headers_in)
        .returns_nothing()
        .execute(self.cfg)
        .await
    }

    /// Partially set the additional headers to be sent with the webhook.
    pub async fn patch_headers(
        &self,
        app_id: String,
        endpoint_id: String,
        endpoint_headers_patch_in: EndpointHeadersPatchIn,
    ) -> Result<()> {
        crate::request::Request::new(
            http1::Method::PATCH,
            "/api/v1/app/{app_id}/endpoint/{endpoint_id}/headers",
        )
        .with_path_param(
            "app_id", app_id,
        )
        .with_path_param(
            "endpoint_id",
            endpoint_id,
        )
        .with_body_param(endpoint_headers_patch_in)
        .returns_nothing()
        .execute(self.cfg)
        .await
    }

    /// Resend all failed messages since a given time.
    ///
    /// Messages that were sent successfully, even if failed initially, are not
    /// resent.
    ///
    /// A completed task will return a payload like the following:
    /// ```json
    /// {
    ///   "id": "qtask_33qen93MNuelBAq1T9G7eHLJRsF",
    ///   "status": "finished",
    ///   "task": "endpoint.recover",
    ///   "data": {
    ///     "messagesSent": 2
    ///   }
    /// }
    /// ```
    pub async fn recover(
        &self,
        app_id: String,
        endpoint_id: String,
        recover_in: RecoverIn,
        options: Option<EndpointRecoverOptions>,
    ) -> Result<RecoverOut> {
        let EndpointRecoverOptions {
            idempotency_key,
        } = options.unwrap_or_default();

        crate::request::Request::new(
            http1::Method::POST,
            "/api/v1/app/{app_id}/endpoint/{endpoint_id}/recover",
        )
        .with_path_param(
            "app_id", app_id,
        )
        .with_path_param(
            "endpoint_id",
            endpoint_id,
        )
        .with_optional_header_param(
            "idempotency-key",
            idempotency_key,
        )
        .with_body_param(recover_in)
        .execute(self.cfg)
        .await
    }

    /// Replays messages to the endpoint.
    ///
    /// Only messages that were created after `since` will be sent.
    /// Messages that were previously sent to the endpoint are not resent.
    ///
    /// A completed task will return a payload like the following:
    /// ```json
    /// {
    ///   "id": "qtask_33qen93MNuelBAq1T9G7eHLJRsF",
    ///   "status": "finished",
    ///   "task": "endpoint.replay",
    ///   "data": {
    ///     "messagesSent": 2
    ///   }
    /// }
    /// ```
    pub async fn replay_missing(
        &self,
        app_id: String,
        endpoint_id: String,
        replay_in: ReplayIn,
        options: Option<EndpointReplayMissingOptions>,
    ) -> Result<ReplayOut> {
        let EndpointReplayMissingOptions {
            idempotency_key,
        } = options.unwrap_or_default();

        crate::request::Request::new(
            http1::Method::POST,
            "/api/v1/app/{app_id}/endpoint/{endpoint_id}/replay-missing",
        )
        .with_path_param(
            "app_id", app_id,
        )
        .with_path_param(
            "endpoint_id",
            endpoint_id,
        )
        .with_optional_header_param(
            "idempotency-key",
            idempotency_key,
        )
        .with_body_param(replay_in)
        .execute(self.cfg)
        .await
    }

    /// Get the endpoint's signing secret.
    ///
    /// This is used to verify the authenticity of the webhook.
    /// For more information please refer to [the consuming webhooks docs](https://docs.svix.com/consuming-webhooks/).
    pub async fn get_secret(
        &self,
        app_id: String,
        endpoint_id: String,
    ) -> Result<EndpointSecretOut> {
        crate::request::Request::new(
            http1::Method::GET,
            "/api/v1/app/{app_id}/endpoint/{endpoint_id}/secret",
        )
        .with_path_param(
            "app_id", app_id,
        )
        .with_path_param(
            "endpoint_id",
            endpoint_id,
        )
        .execute(self.cfg)
        .await
    }

    /// Rotates the endpoint's signing secret.
    ///
    /// The previous secret will remain valid for the next 24 hours.
    pub async fn rotate_secret(
        &self,
        app_id: String,
        endpoint_id: String,
        endpoint_secret_rotate_in: EndpointSecretRotateIn,
        options: Option<EndpointRotateSecretOptions>,
    ) -> Result<()> {
        let EndpointRotateSecretOptions {
            idempotency_key,
        } = options.unwrap_or_default();

        crate::request::Request::new(
            http1::Method::POST,
            "/api/v1/app/{app_id}/endpoint/{endpoint_id}/secret/rotate",
        )
        .with_path_param(
            "app_id", app_id,
        )
        .with_path_param(
            "endpoint_id",
            endpoint_id,
        )
        .with_optional_header_param(
            "idempotency-key",
            idempotency_key,
        )
        .with_body_param(endpoint_secret_rotate_in)
        .returns_nothing()
        .execute(self.cfg)
        .await
    }

    /// Send an example message for an event.
    pub async fn send_example(
        &self,
        app_id: String,
        endpoint_id: String,
        event_example_in: EventExampleIn,
        options: Option<EndpointSendExampleOptions>,
    ) -> Result<MessageOut> {
        let EndpointSendExampleOptions {
            idempotency_key,
        } = options.unwrap_or_default();

        crate::request::Request::new(
            http1::Method::POST,
            "/api/v1/app/{app_id}/endpoint/{endpoint_id}/send-example",
        )
        .with_path_param(
            "app_id", app_id,
        )
        .with_path_param(
            "endpoint_id",
            endpoint_id,
        )
        .with_optional_header_param(
            "idempotency-key",
            idempotency_key,
        )
        .with_body_param(event_example_in)
        .execute(self.cfg)
        .await
    }

    /// Get basic statistics for the endpoint.
    pub async fn get_stats(
        &self,
        app_id: String,
        endpoint_id: String,
        options: Option<EndpointGetStatsOptions>,
    ) -> Result<EndpointStats> {
        let EndpointGetStatsOptions {
            since,
            until,
        } = options.unwrap_or_default();

        crate::request::Request::new(
            http1::Method::GET,
            "/api/v1/app/{app_id}/endpoint/{endpoint_id}/stats",
        )
        .with_path_param(
            "app_id", app_id,
        )
        .with_path_param(
            "endpoint_id",
            endpoint_id,
        )
        .with_optional_query_param(
            "since", since,
        )
        .with_optional_query_param(
            "until", until,
        )
        .execute(self.cfg)
        .await
    }

    /// Get the transformation code associated with this endpoint.
    pub async fn transformation_get(
        &self,
        app_id: String,
        endpoint_id: String,
    ) -> Result<EndpointTransformationOut> {
        crate::request::Request::new(
            http1::Method::GET,
            "/api/v1/app/{app_id}/endpoint/{endpoint_id}/transformation",
        )
        .with_path_param(
            "app_id", app_id,
        )
        .with_path_param(
            "endpoint_id",
            endpoint_id,
        )
        .execute(self.cfg)
        .await
    }

    /// Set or unset the transformation code associated with this endpoint.
    pub async fn patch_transformation(
        &self,
        app_id: String,
        endpoint_id: String,
        endpoint_transformation_patch: EndpointTransformationPatch,
    ) -> Result<()> {
        crate::request::Request::new(
            http1::Method::PATCH,
            "/api/v1/app/{app_id}/endpoint/{endpoint_id}/transformation",
        )
        .with_path_param(
            "app_id", app_id,
        )
        .with_path_param(
            "endpoint_id",
            endpoint_id,
        )
        .with_body_param(endpoint_transformation_patch)
        .returns_nothing()
        .execute(self.cfg)
        .await
    }

    /// This operation was renamed to `set-transformation`.
    #[deprecated]
    pub async fn transformation_partial_update(
        &self,
        app_id: String,
        endpoint_id: String,
        endpoint_transformation_in: EndpointTransformationIn,
    ) -> Result<()> {
        crate::request::Request::new(
            http1::Method::PATCH,
            "/api/v1/app/{app_id}/endpoint/{endpoint_id}/transformation",
        )
        .with_path_param(
            "app_id", app_id,
        )
        .with_path_param(
            "endpoint_id",
            endpoint_id,
        )
        .with_body_param(endpoint_transformation_in)
        .returns_nothing()
        .execute(self.cfg)
        .await
    }
}

File: php/src/Api/Endpoint.php
<?php

// this file is @generated
declare(strict_types=1);

namespace Svix\Api;

use Svix\Models\EndpointHeadersIn;
use Svix\Models\EndpointHeadersOut;
use Svix\Models\EndpointHeadersPatchIn;
use Svix\Models\EndpointIn;
use Svix\Models\EndpointOut;
use Svix\Models\EndpointPatch;
use Svix\Models\EndpointSecretOut;
use Svix\Models\EndpointSecretRotateIn;
use Svix\Models\EndpointStats;
use Svix\Models\EndpointTransformationIn;
use Svix\Models\EndpointTransformationOut;
use Svix\Models\EndpointTransformationPatch;
use Svix\Models\EndpointUpdate;
use Svix\Models\EventExampleIn;
use Svix\Models\ListResponseEndpointOut;
use Svix\Models\MessageOut;
use Svix\Models\RecoverIn;
use Svix\Models\RecoverOut;
use Svix\Models\ReplayIn;
use Svix\Models\ReplayOut;
use Svix\Request\SvixHttpClient;

class Endpoint
{
    public function __construct(
        private readonly SvixHttpClient $client,
    ) {
    }

    /** List the application's endpoints. */
    public function list(
        string $appId,
        ?EndpointListOptions $options = null,
    ): ListResponseEndpointOut {
        $request = $this->client->newReq('GET', "/api/v1/app/{$appId}/endpoint");
        if (null !== $options) {
            $request->setQueryParam('limit', $options->limit);
            $request->setQueryParam('iterator', $options->iterator);
            $request->setQueryParam('order', $options->order);
        }
        $res = $this->client->send($request);

        return ListResponseEndpointOut::fromJson($res);
    }

    /**
     * Create a new endpoint for the application.
     *
     * When `secret` is `null` the secret is automatically generated (recommended).
     */
    public function create(
        string $appId,
        EndpointIn $endpointIn,
        ?EndpointCreateOptions $options = null,
    ): EndpointOut {
        $request = $this->client->newReq('POST', "/api/v1/app/{$appId}/endpoint");
        if (null !== $options) {
            $request->setHeaderParam('idempotency-key', $options->idempotencyKey);
        }
        $request->setBody(json_encode($endpointIn));
        $res = $this->client->send($request);

        return EndpointOut::fromJson($res);
    }

    /** Get an endpoint. */
    public function get(
        string $appId,
        string $endpointId,
    ): EndpointOut {
        $request = $this->client->newReq('GET', "/api/v1/app/{$appId}/endpoint/{$endpointId}");
        $res = $this->client->send($request);

        return EndpointOut::fromJson($res);
    }

    /** Update an endpoint. */
    public function update(
        string $appId,
        string $endpointId,
        EndpointUpdate $endpointUpdate,
    ): EndpointOut {
        $request = $this->client->newReq('PUT', "/api/v1/app/{$appId}/endpoint/{$endpointId}");
        $request->setBody(json_encode($endpointUpdate));
        $res = $this->client->send($request);

        return EndpointOut::fromJson($res);
    }

    /** Delete an endpoint. */
    public function delete(
        string $appId,
        string $endpointId,
    ): void {
        $request = $this->client->newReq('DELETE', "/api/v1/app/{$appId}/endpoint/{$endpointId}");
        $res = $this->client->sendNoResponseBody($request);
    }

    /** Partially update an endpoint. */
    public function patch(
        string $appId,
        string $endpointId,
        EndpointPatch $endpointPatch,
    ): EndpointOut {
        $request = $this->client->newReq('PATCH', "/api/v1/app/{$appId}/endpoint/{$endpointId}");
        $request->setBody(json_encode($endpointPatch));
        $res = $this->client->send($request);

        return EndpointOut::fromJson($res);
    }

    /** Get the additional headers to be sent with the webhook. */
    public function getHeaders(
        string $appId,
        string $endpointId,
    ): EndpointHeadersOut {
        $request = $this->client->newReq('GET', "/api/v1/app/{$appId}/endpoint/{$endpointId}/headers");
        $res = $this->client->send($request);

        return EndpointHeadersOut::fromJson($res);
    }

    /** Set the additional headers to be sent with the webhook. */
    public function updateHeaders(
        string $appId,
        string $endpointId,
        EndpointHeadersIn $endpointHeadersIn,
    ): void {
        $request = $this->client->newReq('PUT', "/api/v1/app/{$appId}/endpoint/{$endpointId}/headers");
        $request->setBody(json_encode($endpointHeadersIn));
        $res = $this->client->sendNoResponseBody($request);
    }

    /** Partially set the additional headers to be sent with the webhook. */
    public function patchHeaders(
        string $appId,
        string $endpointId,
        EndpointHeadersPatchIn $endpointHeadersPatchIn,
    ): void {
        $request = $this->client->newReq('PATCH', "/api/v1/app/{$appId}/endpoint/{$endpointId}/headers");
        $request->setBody(json_encode($endpointHeadersPatchIn));
        $res = $this->client->sendNoResponseBody($request);
    }

    /**
     * Resend all failed messages since a given time.
     *
     * Messages that were sent successfully, even if failed initially, are not resent.
     */
    public function recover(
        string $appId,
        string $endpointId,
        RecoverIn $recoverIn,
        ?EndpointRecoverOptions $options = null,
    ): RecoverOut {
        $request = $this->client->newReq('POST', "/api/v1/app/{$appId}/endpoint/{$endpointId}/recover");
        if (null !== $options) {
            $request->setHeaderParam('idempotency-key', $options->idempotencyKey);
        }
        $request->setBody(json_encode($recoverIn));
        $res = $this->client->send($request);

        return RecoverOut::fromJson($res);
    }

    /**
     * Replays messages to the endpoint.
     *
     * Only messages that were created after `since` will be sent.
     * Messages that were previously sent to the endpoint are not resent.
     */
    public function replayMissing(
        string $appId,
        string $endpointId,
        ReplayIn $replayIn,
        ?EndpointReplayMissingOptions $options = null,
    ): ReplayOut {
        $request = $this->client->newReq('POST', "/api/v1/app/{$appId}/endpoint/{$endpointId}/replay-missing");
        if (null !== $options) {
            $request->setHeaderParam('idempotency-key', $options->idempotencyKey);
        }
        $request->setBody(json_encode($replayIn));
        $res = $this->client->send($request);

        return ReplayOut::fromJson($res);
    }

    /**
     * Get the endpoint's signing secret.
     *
     * This is used to verify the authenticity of the webhook.
     * For more information please refer to [the consuming webhooks docs](https://docs.svix.com/consuming-webhooks/).
     */
    public function getSecret(
        string $appId,
        string $endpointId,
    ): EndpointSecretOut {
        $request = $this->client->newReq('GET', "/api/v1/app/{$appId}/endpoint/{$endpointId}/secret");
        $res = $this->client->send($request);

        return EndpointSecretOut::fromJson($res);
    }

    /**
     * Rotates the endpoint's signing secret.
     *
     * The previous secret will remain valid for the next 24 hours.
     */
    public function rotateSecret(
        string $appId,
        string $endpointId,
        EndpointSecretRotateIn $endpointSecretRotateIn,
        ?EndpointRotateSecretOptions $options = null,
    ): void {
        $request = $this->client->newReq('POST', "/api/v1/app/{$appId}/endpoint/{$endpointId}/secret/rotate");
        if (null !== $options) {
            $request->setHeaderParam('idempotency-key', $options->idempotencyKey);
        }
        $request->setBody(json_encode($endpointSecretRotateIn));
        $res = $this->client->sendNoResponseBody($request);
    }

    /** Send an example message for an event. */
    public function sendExample(
        string $appId,
        string $endpointId,
        EventExampleIn $eventExampleIn,
        ?EndpointSendExampleOptions $options = null,
    ): MessageOut {
        $request = $this->client->newReq('POST', "/api/v1/app/{$appId}/endpoint/{$endpointId}/send-example");
        if (null !== $options) {
            $request->setHeaderParam('idempotency-key', $options->idempotencyKey);
        }
        $request->setBody(json_encode($eventExampleIn));
        $res = $this->client->send($request);

        return MessageOut::fromJson($res);
    }

    /** Get basic statistics for the endpoint. */
    public function getStats(
        string $appId,
        string $endpointId,
        ?EndpointGetStatsOptions $options = null,
    ): EndpointStats {
        $request = $this->client->newReq('GET', "/api/v1/app/{$appId}/endpoint/{$endpointId}/stats");
        if (null !== $options) {
            $request->setQueryParam('since', $options->since);
            $request->setQueryParam('until', $options->until);
        }
        $res = $this->client->send($request);

        return EndpointStats::fromJson($res);
    }

    /** Get the transformation code associated with this endpoint. */
    public function transformationGet(
        string $appId,
        string $endpointId,
    ): EndpointTransformationOut {
        $request = $this->client->newReq('GET', "/api/v1/app/{$appId}/endpoint/{$endpointId}/transformation");
        $res = $this->client->send($request);

        return EndpointTransformationOut::fromJson($res);
    }

    /** Set or unset the transformation code associated with this endpoint. */
    public function patchTransformation(
        string $appId,
        string $endpointId,
        EndpointTransformationPatch $endpointTransformationPatch,
    ): void {
        $request = $this->client->newReq('PATCH', "/api/v1/app/{$appId}/endpoint/{$endpointId}/transformation");
        $request->setBody(json_encode($endpointTransformationPatch));
        $res = $this->client->sendNoResponseBody($request);
    }

    /** This operation was renamed to `set-transformation`. */
    public function transformationPartialUpdate(
        string $appId,
        string $endpointId,
        EndpointTransformationIn $endpointTransformationIn,
    ): void {
        $request = $this->client->newReq('PATCH', "/api/v1/app/{$appId}/endpoint/{$endpointId}/transformation");
        $request->setBody(json_encode($endpointTransformationIn));
        $res = $this->client->sendNoResponseBody($request);
    }
}
